### PR TITLE
Collect column statistics during write

### PR DESCRIPTION
**Pull request by kewang1024**
Prestissimo change: https://github.com/prestodb/presto/pull/20238

### CODE REVIEW

File: velox/exec/TableWriter.h [hunk 1]
```diff
@@ -78,5 +83,8 @@ class TableWriter : public Operator {
   bool finished_ = false;
   bool closed_ = false;
   vector_size_t numWrittenRows_ = 0;
+
+  const static int CONTEXT_CHANNEL = 2;
```
Comment (xiaoxmeng) [inline velox/exec/TableWriter.h:87]: Use the consts defined in TableWriterTraits

File: velox/exec/TableWriter.cpp [hunk 2]
```diff
@@ -91,13 +96,55 @@ void TableWriter::addInput(RowVectorPtr input) {
   }
   dataSink_->appendData(mappedInput);
   numWrittenRows_ += input->size();
+  if (aggregation_ != nullptr) {
+    aggregation_->addInput(input);
+  }
 }
 
+/// Both the statistics and the row_count + fragments are transferred over the
```
Comment (xiaoxmeng) [inline velox/exec/TableWriter.cpp:104]: Nice comment. Could you move to TableWriterTraits. Also comment in cc starting with '//'.

**Review by xiaoxmeng — COMMENTED** [review]
@kewang1024 left some comments. Thanks!

File: velox/exec/TableWriter.cpp [hunk 3]
```diff
@@ -155,6 +179,68 @@ RowVectorPtr TableWriter::getOutput() {
...
+// null       null          X          X          X
+// null       null          X          X          X
+RowVectorPtr TableWriter::createStatisticsRowVector(
+    RowVectorPtr aggregationOutput) {
+  vector_size_t numOutputRows = 1;
+  std::vector<VectorPtr> columns;
+
+  // Set commitcontext column.
+  // clang-format off
+  auto commitContextJson = folly::toJson(
```
Comment (xiaoxmeng) [inline velox/exec/TableWriter.cpp:210]: Can you provide a utility say
```
createCommitContextJson(bool lastPage)?
```

File: velox/exec/TableWriter.h [hunk 4]
```diff
@@ -26,6 +26,29 @@ namespace facebook::velox::exec {
...
+  ///
+  /// For convenience, we never set both: [row_count_channel] +
+  /// [fragment_channel] and the [statistic_channel_1] ...
+  /// [statistic_channel_N].
+  ///
+  /// If this is a row that holds statistics - the [row_count_channel] +
+  /// [fragment_channel] will be NULL.
+  ///
+  /// If this is a row that holds the row count
+  /// or the fragment - all the statistics channels will be set to NULL.
```
Comment (xiaoxmeng) [inline velox/exec/TableWriter.h:51]:
```
s/kStatsChannel/kFirstStatsChannel/
```

File: velox/exec/TableWriter.h [hunk 5]
```diff
@@ -26,6 +26,29 @@ namespace facebook::velox::exec {
 class TableWriterTraits {
  public:
   /// Defines the column channels in table writer output.
+  /// Both the statistics and the row_count + fragments are transferred over the
```
Comment (xiaoxmeng) [inline velox/exec/TableWriter.h:29]: Can you rephrase the comment
```
(1) output layout
(2) row count + fragment and stats are not both set? Explain that we first yield the aggregation stats?
```

**Review by xiaoxmeng — COMMENTED** [review]
@kewang1024 thanks for the change!

File: velox/exec/TableWriter.cpp [hunk 6]
```diff
@@ -91,13 +96,27 @@ void TableWriter::addInput(RowVectorPtr input) {
...
+  }
 }
 
 RowVectorPtr TableWriter::getOutput() {
   // Making sure the output is read only once after the write is fully done.
   if (!noMoreInput_ || finished_) {
     return nullptr;
   }
+
+  if (aggregation_ != nullptr && !aggregation_->isFinished()) {
```
[the remark this answers is not in the document]
Comment (kewang1024) [reply velox/exec/TableWriter.cpp:110]: 1. aggregation_->isFinished() should be able to indicate if we have aggregation stats output or not
2. For TableWriter, it's the partial aggregation here

File: velox/exec/tests/TableWriteTest.cpp [hunk 7]
```diff
@@ -1745,6 +1746,150 @@ TEST_P(AllTableWriterTest, tableWriteOutputCheck) {
 
 // TODO: add partitioned table write update mode tests and more failure tests.
 
+TEST_P(AllTableWriterTest, testColumnStatsForWriter) {
```
[the remark this answers is not in the document]
Comment (kewang1024) [reply velox/exec/tests/TableWriteTest.cpp:1761]: I'm leaning towards keeping it, because `tableWriteOutputCheck` can test the plan without column stats aggregation, and `testColumnStatsForWriter ` can test plan with column stats aggregation

the stats collection feature can be turned off  by config and session property

File: velox/exec/TableWriter.h (hunk 5 above)
Comment (kewang1024) [reply velox/exec/TableWriter.h:29]: Actually those are already in the comment in the later paragraph

Comment (xiaoxmeng) [reply velox/exec/TableWriter.h:29]: Can you first comment on the layout and then describe the special treatments on the output?

File: velox/exec/TableWriter.h [hunk 8]
```diff
@@ -86,12 +112,17 @@ class TableWriter : public Operator {
 
  private:
   void createDataSink();
+  RowVectorPtr createAggregationStatsOutput();
```
Comment (xiaoxmeng) [inline velox/exec/TableWriter.h:115]: Leave an empty line between method with comments.

File: velox/exec/tests/TableWriteTest.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (xiaoxmeng) [reply velox/exec/tests/TableWriteTest.cpp:1761]: Yes, let's have a single test with different test options? I think column stats is one test parameter?

File: velox/exec/TableWriter.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (xiaoxmeng) [reply velox/exec/TableWriter.cpp:131]: I think table writer can start yielding result before noMoreInput_ is set?

File: velox/exec/tests/TableWriteTest.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (kewang1024) [reply velox/exec/tests/TableWriteTest.cpp:1761]: The test plan would be different, then the test result would be different as well (column aggregation would have more), I tried to move, but then eventually there's not much to reuse and we would have two tests in one test function which doesn't seem straightforward.

File: velox/exec/tests/utils/PlanBuilder.h [hunk 9]
```diff
@@ -270,7 +273,9 @@ class PlanBuilder {
           connector::CommitStrategy::kNoCommit);
 
   /// Add a TableWriteMergeNode.
-  PlanBuilder& tableWriteMerge();
+  PlanBuilder& tableWriteMerge(
+      RowTypePtr outputType = nullptr,
+      std::shared_ptr<core::AggregationNode> aggregationNode = nullptr);
```
Comment (xiaoxmeng) [inline velox/exec/tests/utils/PlanBuilder.h:278]: I think we can generate outputType based on aggregationNode?

File: velox/exec/tests/utils/PlanBuilder.h [hunk 10]
```diff
@@ -240,14 +240,17 @@ class PlanBuilder {
...
   PlanBuilder& tableWrite(
       const RowTypePtr& inputColumns,
       const std::vector<std::string>& tableColumnNames,
       const std::shared_ptr<core::AggregationNode>& aggregationNode,
       const std::shared_ptr<core::InsertTableHandle>& insertHandle,
       bool hasPartitioningScheme,
       connector::CommitStrategy commitStrategy =
-          connector::CommitStrategy::kNoCommit);
+          connector::CommitStrategy::kNoCommit,
+      RowTypePtr outputType = nullptr);
```
Comment (xiaoxmeng) [inline velox/exec/tests/utils/PlanBuilder.h:253]: I think we can generate outputType based on aggregationNode?

File: velox/exec/tests/utils/PlanBuilder.cpp [hunk 11]
```diff
@@ -303,23 +303,30 @@ PlanBuilder& PlanBuilder::tableWrite(
...
+    RowTypePtr outputType) {
   planNode_ = std::make_shared<core::TableWriteNode>(
       nextPlanNodeId(),
       inputColumns,
       tableColumnNames,
       aggregationNode,
       insertHandle,
       hasPartitioningScheme,
-      TableWriteTraits::outputType(),
+      outputType == nullptr ? TableWriteTraits::outputType() : outputType,
```
Comment (xiaoxmeng) [inline velox/exec/tests/utils/PlanBuilder.cpp:315]: Can we extend outputType() take an optional aggregationNode to generate the output type?

File: velox/exec/TableWriter.cpp [hunk 12]
```diff
@@ -134,36 +159,88 @@ RowVectorPtr TableWriter::getOutput() {
...
+  for (int channel = 0; channel < outputType->size(); channel++) {
+    if (channel < TableWriteTraits::kContextChannel) {
+      // 1. Set null rows column.
+      // 2. Set null fragments column.
+      auto nullRowCountOrFragmentVector =
+          BaseVector::create(outputType->childAt(channel), numOutputRows, pool);
+      for (int row = 0; row < numOutputRows; row++) {
+        nullRowCountOrFragmentVector->setNull(row, true);
+      }
+      columns.push_back(nullRowCountOrFragmentVector);
```
Comment (xiaoxmeng) [inline velox/exec/TableWriter.cpp:225]: nit: use pattern
```
if () {
  continue;
}
if () {
  continue;
}
...
```

**Review by xiaoxmeng — COMMENTED** [review]
@kewang1024 Looks pretty good % minors. Haven't looked at the test yet. Thanks!

File: velox/exec/TableWriter.h [hunk 13]
```diff
@@ -101,17 +127,31 @@ class TableWriter : public Operator {
     return finished_;
   }
 
+  // Creates the statistics output.
+  // Statistics page layout (aggregate by partition):
+  // row     fragments     context     stats1     stats2 ...
+  // null       null          X          X          X
+  // null       null          X          X          X
+  static RowVectorPtr createAggregationStatsOutput(
```
[the remark this answers is not in the document]
Comment (kewang1024) [reply velox/exec/TableWriter.h:135]: It is used by the table writer merge operator

File: velox/exec/TableWriteMerge.cpp [hunk 14]
```diff
@@ -48,20 +49,28 @@ bool containsNonNullRows(const VectorPtr& vector) {
...
+    aggregation_ = std::make_unique<HashAggregation>(
+        operatorId, driverCtx, tableWriteMergeNode->aggregationNode());
+  }
 }
 
 void TableWriteMerge::addInput(RowVectorPtr input) {
   VELOX_CHECK(!noMoreInput_);
   VELOX_CHECK_GT(input->size(), 0);
 
+  if (aggregation_ != nullptr && isStatistics(input)) {
```
[the remark this answers is not in the document]
Comment (kewang1024) [reply velox/exec/TableWriteMerge.cpp:69]: So this function is to see if the input is statistics result so I felt `isStatistics` is a better name

File: velox/exec/TableWriter.cpp [hunk 15]
```diff
@@ -134,36 +159,88 @@ RowVectorPtr TableWriter::getOutput() {
...
 }
 
+// static
+RowVectorPtr TableWriter::createAggregationStatsOutput(
+    RowTypePtr outputType,
+    RowVectorPtr aggregationOutput,
+    StringView tableCommitContext,
+    velox::memory::MemoryPool* pool) {
+  // TODO: record aggregation stats output time.
+  if (aggregationOutput == nullptr || aggregationOutput->childrenSize() == 0) {
```
[the remark this answers is not in the document]
Comment (kewang1024) [reply velox/exec/TableWriter.cpp:211]: then we need to pass in the whole aggregation operator inside, but in this function, we only need its aggregation output, so it would be better to only pass in aggregationOutput

File: velox/exec/tests/utils/PlanBuilder.h (hunk 9 above)
Comment (kewang1024) [reply velox/exec/tests/utils/PlanBuilder.h:278]: Yeah, but the outputType from TableWriter is originally generated from aggregationNode from upper level (presto): https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp#L2139-L2159

So in lower level (velox), generating again feels redundant to me and I think we can directly reuse

File: velox/exec/TableWriter.cpp [hunk 16]
```diff
@@ -134,36 +171,86 @@ RowVectorPtr TableWriter::getOutput() {
...
   lockedStats->physicalWrittenBytes = writtenBytes;
 }
 
+// static
+RowVectorPtr TableWriteTraits::createAggregationStatsOutput(
+    RowTypePtr outputType,
+    RowVectorPtr aggregationOutput,
+    StringView tableCommitContext,
+    velox::memory::MemoryPool* pool) {
+  // TODO: record aggregation stats output time.
```
Comment (xiaoxmeng) [inline velox/exec/TableWriter.cpp:222]: Can we enforce the check that aggregationOutput can't be null and childrenSize can be zero?

File: velox/exec/TableWriter.cpp [hunk 17]
```diff
@@ -134,36 +171,86 @@ RowVectorPtr TableWriter::getOutput() {
...
+      continue;
+    }
+    if (channel == TableWriteTraits::kContextChannel) {
+      // 3. Set commitcontext column.
+      columns.push_back(std::make_shared<ConstantVector<StringView>>(
+          pool,
+          numOutputRows,
+          false /*isNull*/,
+          VARBINARY(),
+          std::move(tableCommitContext)));
```
Comment (xiaoxmeng) [inline velox/exec/TableWriter.cpp:243]: you don't need to move a stringView and let's just copy and it is a bit fragile doing so in a loop if call this more than once.

File: velox/exec/tests/TableWriteTest.cpp [hunk 18]
```diff
@@ -2152,6 +2156,251 @@ TEST_P(AllTableWriterTest, tableWriteOutputCheck) {
...
+  std::vector<core::FieldAccessTypedExprPtr> groupingKeys;
+  // 2. partition columns
+  for (int i = 0; i < partitionedBy_.size(); i++) {
+    groupingKeys.emplace_back(
+        std::make_shared<const core::FieldAccessTypedExpr>(
+            partitionTypes_.at(i), partitionedBy_.at(i)));
+    output.emplace_back(partitionedBy_.at(i));
+    types.emplace_back(partitionTypes_.at(i));
+  }
+  // 3. stats columns
```
Comment (xiaoxmeng) [inline velox/exec/tests/TableWriteTest.cpp:2177]: Can you put the logic that build aggregationNode into a test utility?

File: velox/exec/tests/TableWriteTest.cpp [hunk 19]
```diff
@@ -2152,6 +2156,251 @@ TEST_P(AllTableWriterTest, tableWriteOutputCheck) {
...
+  //
+  // For unpartitioned, expected result is:
+  // Row     Fragment           Context       partition           c1_min_value
+  // null    null                x                                0
+  // count   null                x            null                null
+  // null    update              x            null                null
+
+  int statsRow = 0;
+  while (columnStatsVector->isNullAt(statsRow) && statsRow < result->size()) {
+    statsRow++;
```
Comment (xiaoxmeng) [inline velox/exec/tests/TableWriteTest.cpp:2385]: nit ++statsRow;

File: velox/exec/tests/TableWriteTest.cpp [hunk 20]
```diff
@@ -2152,6 +2156,251 @@ TEST_P(AllTableWriterTest, tableWriteOutputCheck) {
...
+      ASSERT_TRUE(fragmentVector->isNullAt(i));
+    } else {
+      ASSERT_TRUE(columnStatsVector->isNullAt(i));
+      ASSERT_TRUE(rowVector->isNullAt(i));
+      ASSERT_FALSE(fragmentVector->isNullAt(i));
+    }
+  }
+}
+
+TEST_P(AllTableWriterTest, columnStatsTableWriteMerge) {
```
Comment (xiaoxmeng) [inline velox/exec/tests/TableWriteTest.cpp:2272]:
```
s/columnStatsTableWriteMerge/columnStatsWithTableWriteMerge/
```

File: velox/exec/tests/TableWriteTest.cpp [hunk 21]
```diff
@@ -2152,6 +2156,251 @@ TEST_P(AllTableWriterTest, tableWriteOutputCheck) {
...
+      ASSERT_TRUE(fragmentVector->isNullAt(i));
+      ASSERT_FALSE(columnStatsVector->isNullAt(i));
+    } else {
+      ASSERT_FALSE(rowVector->isNullAt(i));
+      ASSERT_TRUE(fragmentVector->isNullAt(i));
+      ASSERT_TRUE(columnStatsVector->isNullAt(i));
+    }
+  }
+}
+
```
Comment (xiaoxmeng) [inline velox/exec/tests/TableWriteTest.cpp:2403]: Can you add tests in followup to verify the each type of stats are correctly generated? Thanks!

File: velox/type/Type.h [hunk 22]
```diff
@@ -875,7 +875,8 @@ class RowType : public TypeBase<TypeKind::ROW> {
   void printChildren(std::stringstream& ss, std::string_view delimiter = ",")
       const;
 
-  std::shared_ptr<RowType> unionWith(std::shared_ptr<RowType>& rowType) const;
+  std::shared_ptr<RowType> unionWith(
```
Comment (xiaoxmeng) [inline velox/type/Type.h:878]:
```
const std::shared_ptr<const RowType>& rowType
```
Can you add a unit test for this? Thanks!

File: velox/exec/TableWriter.h [hunk 23]
```diff
@@ -51,14 +74,25 @@ class TableWriteTraits {
       "pageSinkCommitStrategy";
   static constexpr std::string_view klastPageContextKey = "lastPage";
 
-  /// TODO: add column stats support.
-  static const RowTypePtr& outputType();
+  static const RowTypePtr outputType(
+      std::shared_ptr<core::AggregationNode> aggregationNode = nullptr);
```
Comment (xiaoxmeng) [inline velox/exec/TableWriter.h:78]:
```
const std::shared_ptr<core::AggregationNode>&
```

File: velox/exec/TableWriter.cpp [hunk 24]
```diff
@@ -134,36 +166,87 @@ RowVectorPtr TableWriter::getOutput() {
...
-      StringView(commitContextJson));
+      StringView(commitContext));
 
   std::vector<VectorPtr> columns = {
       writtenRowsVector, fragmentsVector, commitContextVector};
 
+  // 4. Set null statistics columns.
+  if (aggregation_ != nullptr) {
+    for (int i = TableWriteTraits::kStatsChannel; i < outputType_->size();
+         i++) {
```
Comment (xiaoxmeng) [inline velox/exec/TableWriter.cpp:184]:
```
nit: s/i++/++i/
```

File: velox/exec/TableWriter.cpp [hunk 25]
```diff
@@ -134,36 +166,87 @@ RowVectorPtr TableWriter::getOutput() {
...
+RowVectorPtr TableWriteTraits::createAggregationStatsOutput(
+    RowTypePtr outputType,
+    RowVectorPtr aggregationOutput,
+    StringView tableCommitContext,
+    velox::memory::MemoryPool* pool) {
+  // TODO: record aggregation stats output time.
+  if (aggregationOutput == nullptr) {
+    return nullptr;
+  }
+  VELOX_CHECK_NE(aggregationOutput->childrenSize(), 0);
```
Comment (xiaoxmeng) [inline velox/exec/TableWriter.cpp:221]:
```
s/VELOX_CHECK_NE/VELOX_CHECK_GT/
```

File: velox/exec/TableWriteMerge.cpp [hunk 26]
```diff
@@ -48,20 +49,31 @@ bool containsNonNullRows(const VectorPtr& vector) {
...
     : Operator(
           driverCtx,
-          TableWriteMergeNode->outputType(),
+          tableWriteMergeNode->outputType(),
           operatorId,
-          TableWriteMergeNode->id(),
+          tableWriteMergeNode->id(),
           "TableWriteMerge") {
-  VELOX_CHECK(outputType_->equivalent(*TableWriteTraits::outputType()));
+  VELOX_CHECK(outputType_->equivalent(
```
Comment (xiaoxmeng) [inline velox/exec/TableWriteMerge.cpp:59]:
```
NYC: s/VELOX_CHECK/VELOX_USER_CHECK/
```

File: velox/exec/TableWriter.cpp [hunk 27]
```diff
@@ -40,6 +41,17 @@ TableWriter::TableWriter(
       insertTableHandle_(
           tableWriteNode->insertTableHandle()->connectorInsertTableHandle()),
       commitStrategy_(tableWriteNode->commitStrategy()) {
+  if (tableWriteNode->outputType()->size() == 1) {
+    VELOX_USER_CHECK_NULL(tableWriteNode->aggregationNode());
+  } else {
+    tableWriteNode->outputType()->equivalent(
```
Comment (xiaoxmeng) [inline velox/exec/TableWriter.cpp:47]:
```
VELOX_USER_CHECK()
```

File: velox/exec/TableWriter.cpp [hunk 28]
```diff
@@ -194,11 +277,16 @@ const TypePtr& TableWriteTraits::contextColumnType() {
...
 
-const RowTypePtr& TableWriteTraits::outputType() {
-  static const auto kOutputType =
+const RowTypePtr TableWriteTraits::outputType(
+    std::shared_ptr<core::AggregationNode> aggregationNode) {
+  auto aggregationOutputType = ROW({}, {});
+  if (aggregationNode != nullptr) {
+    aggregationOutputType = aggregationNode->outputType();
+  }
+  std::shared_ptr<const RowType> kOutputType =
```
Comment (xiaoxmeng) [inline velox/exec/TableWriter.cpp:286]:
```
static const auto kOutputTypeWithoutStats = ;
if (aggregationNode == nullptr) {
  return kOutputTypeWithoutStats;
}
return kOutputTypeWithoutStats->unionWith(aggregationNode->outputType());
```

File: velox/exec/tests/TableWriteTest.cpp [hunk 29]
```diff
@@ -54,6 +55,29 @@ std::string testModeString(TestMode mode) {
   }
 }
 
+static std::shared_ptr<core::AggregationNode> generateAggregationNode(
+    std::string name,
+    std::vector<core::FieldAccessTypedExprPtr> groupingKeys,
```
Comment (xiaoxmeng) [inline velox/exec/tests/TableWriteTest.cpp:60]:
```
const std::vector<core::FieldAccessTypedExprPtr>& groupingKeys
```

File: velox/exec/tests/TableWriteTest.cpp [hunk 30]
```diff
@@ -54,6 +55,29 @@ std::string testModeString(TestMode mode) {
   }
 }
 
+static std::shared_ptr<core::AggregationNode> generateAggregationNode(
+    std::string name,
```
Comment (xiaoxmeng) [inline velox/exec/tests/TableWriteTest.cpp:59]:
```
const std::string& name
```

File: velox/exec/tests/TableWriteTest.cpp [hunk 31]
```diff
@@ -54,6 +55,29 @@ std::string testModeString(TestMode mode) {
...
+    std::vector<core::FieldAccessTypedExprPtr> groupingKeys,
+    AggregationNode::Step step,
+    PlanNodePtr source) {
```
Comment (xiaoxmeng) [inline velox/exec/tests/TableWriteTest.cpp:62]:
```
const PlanNodePtr& source
```

File: velox/exec/tests/TableWriteTest.cpp [hunk 32]
```diff
@@ -54,6 +55,29 @@ std::string testModeString(TestMode mode) {
...
+  core::TypedExprPtr inputField =
+      std::make_shared<const core::FieldAccessTypedExpr>(BIGINT(), name);
+  auto callExpr = std::make_shared<const core::CallTypedExpr>(
+      BIGINT(), std::vector<core::TypedExprPtr>{inputField}, "min");
```
Comment (xiaoxmeng) [inline velox/exec/tests/TableWriteTest.cpp:66]: We only test min for now?

File: velox/exec/tests/utils/PlanBuilder.h (hunk 10 above)
Comment (xiaoxmeng) [inline velox/exec/tests/utils/PlanBuilder.h:253]: We don't need outputType now?

File: velox/exec/tests/utils/PlanBuilder.h [hunk 33]
```diff
@@ -240,14 +240,17 @@ class PlanBuilder {
   /// @param hasPartitioningScheme indicates if table partitioning scheme is
   /// required for this table write which is only true for bucketed hive table
   /// for now.
+  /// @param outputType The output columns, if not specified, use the default:
```
Comment (xiaoxmeng) [inline velox/exec/tests/utils/PlanBuilder.h:243]: Remove this comment?

File: velox/exec/tests/utils/PlanBuilder.h [hunk 34]
```diff
@@ -270,7 +273,9 @@ class PlanBuilder {
           connector::CommitStrategy::kNoCommit);
 
   /// Add a TableWriteMergeNode.
-  PlanBuilder& tableWriteMerge();
+  PlanBuilder& tableWriteMerge(
+      RowTypePtr outputType = nullptr,
```
Comment (xiaoxmeng) [inline velox/exec/tests/utils/PlanBuilder.h:277]: We don't need outputType?

File: velox/exec/tests/utils/PlanBuilder.cpp [hunk 35]
```diff
@@ -303,23 +303,30 @@ PlanBuilder& PlanBuilder::tableWrite(
...
 }
 
-PlanBuilder& PlanBuilder::tableWriteMerge() {
+PlanBuilder& PlanBuilder::tableWriteMerge(
+    RowTypePtr outputType,
+    std::shared_ptr<core::AggregationNode> aggregationNode) {
   planNode_ = std::make_shared<core::TableWriteMergeNode>(
-      nextPlanNodeId(), TableWriteTraits::outputType(), planNode_);
+      nextPlanNodeId(),
+      outputType == nullptr ? TableWriteTraits::outputType()
```
Comment (xiaoxmeng) [inline velox/exec/tests/utils/PlanBuilder.cpp:326]:
```
TableWriteTraits::outputType(aggregationNode)
```

File: velox/exec/tests/utils/PlanBuilder.cpp (hunk 11 above)
Comment (xiaoxmeng) [inline velox/exec/tests/utils/PlanBuilder.cpp:315]: ditto

**Review by xiaoxmeng — COMMENTED** [review]
@kewang1024 thanks for the update!

File: velox/exec/tests/TableWriteTest.cpp (hunk 32 above)
Comment (kewang1024) [reply velox/exec/tests/TableWriteTest.cpp:66]: Yes, for now we only focus on the structure test, in the followup we can start testing for different functions
